lib.nix

```diff
@@ -79,35 +79,42 @@ rec
     , cargolock
     }:
       let
-        tomlDependencies = cargotoml:
-          lib.filter (x: ! isNull x) (
+      tomlDependencies = cargotoml:
+        lib.filter (x: ! isNull x) (
           lib.mapAttrsToList
             (k: v:
               if ! (lib.isAttrs v && builtins.hasAttr "git" v)
               then null
-              else lib.filterAttrs (n: _: n == "rev" || n == "tag" || n == "branch") v //
-                { name = k;
-                  url = v.git;
-                  key = v.rev or v.tag or v.branch or
-                        (throw "No 'rev', 'tag' or 'branch' available to specify key");
+              else
+                let
+                  query = p: p.name == k && (lib.substring 0 (4 + lib.stringLength v.git) p.source) == "git+${v.git}";
+                  extractRevision = url: lib.last (lib.splitString "#" url);
+                  parseLock = lock: rec { inherit (lock) name source; revision = extractRevision source; };
+                  packageLocks = builtins.map parseLock (lib.filter query cargolock.package);
+                  matchByName = lib.findFirst (p: p.name == k) null packageLocks;
+                  matchByRev =
+                    if (isNull (v.rev or null))
+                    then null
+                    else lib.findFirst (p: lib.substring 0 7 p.revision == lib.substring 0 7 v.rev) null packageLocks;
+                  val = if (isNull (v.rev or null)) then v // { rev = matchByRev.revision or matchByName.revision; } else v;
```

Comment:
Here, why go through `matchByRev`? isn't `matchByRev.revision` going to be equal to `v.rev`?

Reply:
Oh yeah, that's right. I only kept that since it was in the original code as well. I can remove that.

Reply:
Alright 10cb38e2704bb1a6cf746f2699da7287ce2e05b0 does that. Tested and it still works correctly on my project.